src/Fhir.Anonymizer.Core/PartitionedExecution/FhirPartitionedExecutor.cs

```diff
@@ -0,0 +1,113 @@
+using System.Threading.Tasks;
+using Fhir.Anonymizer.Core.PartitionedExecution;
+
+namespace Fhir.Anonymizer.Core
+{
+    public class FhirPartitionedExecutor
+    {
+        public IFhirDataReader RawDataReader { set; get; }
+
+        public IFhirDataConsumer AnonymizedDataConsumer { set; get; }
+
+        public Func<string, string> AnonymizerFunction { set; get; }
+
+        public int PartitionCount { set; get; } = Constants.DefaultPartitionedExecutionCount;
+
+        public int BatchSize { set; get; } = Constants.DefaultPartitionedExecutionBatchSize;
+
+        public FhirPartitionedExecutor(IFhirDataReader rawDataReader, IFhirDataConsumer anonymizedDataConsumer, Func<string, string> anonymizerFunction)
+        {
+            RawDataReader = rawDataReader;
+            AnonymizedDataConsumer = anonymizedDataConsumer;
+            AnonymizerFunction = anonymizerFunction;
+        }
+
+        public async Task ExecuteAsync(CancellationToken cancellationToken, bool breakOnAnonymizationException = false, IProgress<BatchAnonymizeResult> progress = null)
+        {
+            Queue<Task<IEnumerable<string>>> executionTasks = new Queue<Task<IEnumerable<string>>>();
+            List<string> batchData = new List<string>();
+
+            string content;
+            while ((content = await RawDataReader.NextAsync().ConfigureAwait(false)) != null)
+            {
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    throw new OperationCanceledException();
+                }
+
+                batchData.Add(content);
+                if (batchData.Count < BatchSize)
+                {
+                    continue;
+                }
+
+                executionTasks.Enqueue(AnonymizeAsync(batchData, breakOnAnonymizationException, progress, cancellationToken));
+                batchData = new List<string>();
+                if (executionTasks.Count < PartitionCount)
+                {
+                    continue;
+                }
+
+                await ConsumeExecutionResultTask(executionTasks).ConfigureAwait(false);
+            }
+
+            if (batchData.Count > 0)
+            {
+                executionTasks.Enqueue(AnonymizeAsync(batchData, breakOnAnonymizationException, progress, cancellationToken));
+            }
+
+            while (executionTasks.Count > 0)
+            {
+                await ConsumeExecutionResultTask(executionTasks).ConfigureAwait(false);
+            }
+
+            await AnonymizedDataConsumer.CompleteAsync().ConfigureAwait(false);
+        }
+
+        private async Task<IEnumerable<string>> AnonymizeAsync(List<string> batchData, bool breakOnAnonymizationException, IProgress<BatchAnonymizeResult> progress, CancellationToken cancellationToken)
+        {
+            return await Task.Run(() =>
+            {
+                List<string> result = new List<string>();
+
+                BatchAnonymizeResult batchResult = new BatchAnonymizeResult();
```

Comment:
The naming of BatchAnonymizeResult looks that it holds the actual anonymized results of this batch, but actually it is more likely the statistic data during processing. Perhaps we could figure out another name for it to distinguish between it and variable in Line 74.

Reply:
Rename to BatchAnonymizeProgressDetail to make it clear to the result. Thanks